### PR TITLE
Implement request cancellation via contexts

### PR DESCRIPTION
**Pull request by mihaitodor**
TODO:
- [x] Do some (manual) testing

### CODE REVIEW

File: faster_raster.go [hunk 1]
```diff
@@ -110,18 +110,20 @@ type Rasterizer struct {
 	stopCompleted      chan struct{}
 }
 
-func NewRasterizer(filename string) *Rasterizer {
+// NewRasterizer returns a configured Rasterizer for a given filename,
+// which is able to process parallelRequests requests simultaneously.
```
Comment (relistan) [inline faster_raster.go]: That's not right. They are not parallel. It's a buffer, as originally named.
```suggestion
// which uses a buffered channel of rasterBufferSize to queue requests for the rasterizer.
```

File: faster_raster.go [hunk 2]
```diff
@@ -110,18 +110,20 @@ type Rasterizer struct {
...
+func NewRasterizer(filename string, parallelRequests int) *Rasterizer {
 	return &Rasterizer{
 		Filename:    filename,
-		RequestChan: make(chan *RasterRequest, RasterBufferSize),
+		RequestChan: make(chan *RasterRequest, parallelRequests),
```
Comment (relistan) [inline faster_raster.go]: Same.

File: faster_raster.go [hunk 3]
```diff
@@ -366,57 +383,85 @@ func (r *Rasterizer) calculateScaleForDocument(pageCount int) {
 	}
 }
 
+func sendErrorReply(req *RasterRequest, filename string, err error) {
```
Comment (relistan) [inline faster_raster.go]: This ought to be a method on the `RasterRequest` IMO

File: faster_raster.go [hunk 4]
```diff
@@ -366,57 +383,85 @@ func (r *Rasterizer) calculateScaleForDocument(pageCount int) {
...
+
+		return nil
+	}
+
+	var cookie C.fz_cookie
+	done := make(chan struct{})
+	go func() {
+		select {
+		case <-req.ctx.Done():
+			cookie.abort = 1
```
Comment (relistan) [inline faster_raster.go]: Is any pthreads locking required around this state modification?

File: faster_raster_test.go [hunk 5]
```diff
@@ -170,10 +171,11 @@ func Test_WithoutFileExtensions(t *testing.T) {
 
 func Test_Processing(t *testing.T) {
 	Convey("When processing the file", t, func() {
-		raster := NewRasterizer("fixtures/sample.pdf")
+		// Allow up to 10 parallel requests
```
Comment (relistan) [inline faster_raster_test.go]: s/parallell/buffered/

File: faster_raster_test.go [hunk 6]
```diff
@@ -204,13 +206,25 @@ func Test_Processing(t *testing.T) {
...
+			defer raster.Stop()
+			So(err, ShouldBeNil)
+
+			ctx, doneFunc := context.WithTimeout(context.Background(), 1*time.Millisecond)
+			defer doneFunc()
+
+			_, err = raster.GeneratePageImage(ctx, 2, 1024, 0)
+			So(err, ShouldEqual, ErrRasterTimeout)
+		})
+
```
Comment (relistan) [inline faster_raster_test.go]: Need a test here where the passed-in context times out as well.

File: faster_raster_test.go [hunk 7]
```diff
@@ -228,6 +242,18 @@ func Test_Processing(t *testing.T) {
 			raster.Stop()
 		})
 
+		Convey("returns an error when the SVG raster request is cancelled", func() {
+			err := raster.Run()
+			defer raster.Stop()
+			So(err, ShouldBeNil)
+
+			ctx, doneFunc := context.WithTimeout(context.Background(), 1*time.Millisecond)
```
Comment (relistan) [inline faster_raster_test.go]: It's really a `cancelFunc` not a `doneFunc`

File: faster_raster_test.go [hunk 8]
```diff
@@ -228,6 +242,18 @@ func Test_Processing(t *testing.T) {
...
+			defer doneFunc()
```
Comment (relistan) [inline faster_raster_test.go]: It's not clear to me that it's necessary to store and call this?

File: faster_raster.go (hunk 4 above)
Comment (mihaitodor) [reply faster_raster.go]: I was wondering the same, but it seems that's not needed. See [here](https://github.com/ArtifexSoftware/mupdf/blob/9b183c560ef9183f7d4ab32af56d8bad0b494f36/include/mupdf/fitz/device.h#L219) and [here](https://github.com/ArtifexSoftware/mupdf/blob/9b183c560ef9183f7d4ab32af56d8bad0b494f36/source/tools/muraster.c#L680) for the way they use it. Thinking about it, it's just a value in memory that they read periodically and we have no control over when this read occurs. If we change the value, they'll get the updated version eventually and bail out early if the processing hasn't finished yet.

File: faster_raster_test.go (hunk 8 above)
Comment (mihaitodor) [reply faster_raster_test.go]: It's not. Doesn't do anything useful in this case. I'm removing it.

File: faster_raster_test.go (hunk 6 above)
Comment (mihaitodor) [reply faster_raster_test.go]: I'll see about adding some more tests in the future and maybe another refactor pass. For now, I'm returning a timeout error even when the context is cancelled, but I should probably return a cancelled error.

**Pull request merged**